homeassistant/core.py

```diff
@@ -1049,9 +1049,17 @@ def async_entity_ids_count(
 
     def all(self, domain_filter: str | Iterable | None = None) -> list[State]:
         """Create a list of all states."""
-        return run_callback_threadsafe(
-            self._loop, self.async_all, domain_filter
-        ).result()
+        if domain_filter is None:
+            return list(self._states.values())
```

Comment:
Are we sure that this is threadsafe.

Reply:
It's absolutely not OK as per the Python spec, but assuming we will always use CPython it's OK.

Reply:
I would not change it.

Reply:
Reverted in b0ae32a2bc14ec4c7e9be437a0b6f21c17a2bac8, we can discuss potential optimizations of `StateMachine` methods separately.